### PR TITLE
PHP-468 custom endpoint on client creation, with unit tests

### PR DESCRIPTION
**Pull request by shababcardinity**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -63,16 +63,7 @@ public static function create(array $options = [], $logger = Client::LOG_NONE)
...
-                new Url(),
-            ]);
-            if(count($violations) != 0){
-                throw new Exception\InvalidAttributeValue(
-                    'Your API endpoint is not a valid URL',
-                    $violations
-                );
-            }
-            Client::$url = $options['apiEndpoint'];
+            Client::validateClientEndpoint($options, $validator);
```
Comment (arturasfio) [inline src/Client.php]: Is there a particular reason why a full class name is used instead of `self` ?

File: src/Client.php [hunk 2]
```diff
@@ -201,4 +192,24 @@ private function prepareAttributes(array $data)
...
+        $violations = $validator->validate($options['apiEndpoint'], [
+            new Url(),
+        ]);
+        if(count($violations) != 0){
+            throw new Exception\InvalidAttributeValue(
+                'Your API endpoint is not a valid URL',
+                $violations
+            );
+        }
+        Client::$url = $options['apiEndpoint'];
```
Comment (arturasfio) [inline src/Client.php]: Assigning the new parameter should not be a responsibility of a method that does validation. I would move this to line 68, after the validator returns without errors

File: tests/ClientEndpointTest.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * @return void
+     */
+    public function testValidURLValidEndpoint(){
+
+        $this->baseConfig['apiEndpoint'] = 'https://api.cardinity.com/v1/';
+        try {
+            $client = Client::create($this->baseConfig, $this->log);
+        } catch (\Throwable $th) {
+            //throw $th;
```
Comment (arturasfio) [inline tests/ClientEndpointTest.php]: What is the reason for silently catching exception?

File: tests/ClientEndpointTest.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * @return void
+     */
+    public function testNoEndpoint(){
+
+        unset($this->baseConfig['apiEndpoint']);
+        try {
+            $client = Client::create($this->baseConfig, $this->log);
+        } catch (\Throwable $th) {
+            //throw $th;
```
Comment (arturasfio) [inline tests/ClientEndpointTest.php]: What is the reason for silently catching exception?

File: tests/ClientEndpointTest.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        }
+        $this->assertInstanceOf('Cardinity\Client', $client);
+    }
+
+     /**
+     * Test create client with valid URL,
+     *
+     * @return void
+     */
+    public function testNoEndpoint(){
```
Comment (arturasfio) [inline tests/ClientEndpointTest.php:73]: If there is no endpoint specified, the default URL should be used, but does this test verify that?

Comment (shababcardinity) [reply tests/ClientEndpointTest.php:73]: what would be a good way to verify that,

File: tests/ClientEndpointTest.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+     * @return void
+     */
+    public function testValidUrlWrongEndpoint(){
+
+        $this->baseConfig['apiEndpoint'] = 'https://example.com/';
+
+        try {
+            $client = Client::create($this->baseConfig, $this->log);
+        } catch (\Exception $e) {
+            //throw $th;
```
Comment (arturasfio) [inline tests/ClientEndpointTest.php]: Redundant comment

File: tests/ClientEndpointTest.php [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+     *
+     * @return void
+     */
+    public function testInvalidURLEndpoint(){
+
+        $this->baseConfig['apiEndpoint'] = '123abc';
+        try {
+            Client::create($this->baseConfig, $this->log);
+        } catch (\Exception $e) {
+            //throw $th;
```
Comment (arturasfio) [inline tests/ClientEndpointTest.php]: Redundant comment

**Review by arturasfio — DISMISSED** [review]

**Review by arturasfio — APPROVED** [review]

**Pull request merged**